src/services/schedule.ts

```diff
@@ -227,29 +214,207 @@ export class ScheduleService implements IScheduleService {
+
+  // if newly creating a tx, autopopulate condition to match the asset being transferred
+  if (!transactionExists && isProxyRequest) {
+    transaction.conditionAsset = metadata.assetContract;
+    transaction.conditionAmount = metadata.assetAmountWei;
+    transaction.conditionAssetDecimals = metadata.assetDecimals;
+  }
+
+  return transaction;
+}
+
+// TODO: optimize this function - takes too much time
+async function getTransactionMetadata(transaction: IScheduled): Promise<ITransactionMetadata> {
+  switch (transaction.assetType) {
+    case AssetType.Ethereum:
+    case undefined: {
+      return ethUtils.fetchTransactionMetadata(transaction);
     }
+    case AssetType.Polkadot: {
+      const api = await getApi(transaction.chainId);
+      return api.fetchTransactionMetadata(transaction);
+    }
+  }
+}
+
+async function matchStrategyPrep(transaction: IScheduled): Promise<{ transaction: IScheduled; isStrategyTx: boolean }> {
+  const defaultResult = { transaction, isStrategyTx: false };
+
+  if (!transaction.userId) {
+    return defaultResult;
   }
 
-  private findBySignedTransaction(signedTransaction: string) {
-    return Scheduled.findOne({ signedTransaction }).exec();
+  const userId = transaction.userId!;
+  const hasAnyPrep = await strategyService.hasAnyPrep(userId);
+
+  if (!hasAnyPrep) {
+    return defaultResult;
   }
 
-  private isDevTx(email: string): boolean {
-    return DEV_PAYMENT_EMAILS.includes(email.toLowerCase());
+  const prepTx = decodeTxForStrategyPrep(transaction);
+  const strategyPrep = await strategyService.matchPrep(userId, prepTx);
+  const isStrategyTx = !!strategyPrep;
+
+  if (hasAnyPrep && !isStrategyTx) {
+    throw new Error('User is already executing another strategy');
   }
 
-  private isValidCouponCode(paymentRefundAddress: string): boolean {
-    return COUPON_CODES.includes(paymentRefundAddress.toLowerCase());
+  if (!isStrategyTx) {
+    return defaultResult;
+  }
+
+  transaction.priority = strategyPrep.priority;
+  transaction.conditionAsset = strategyPrep.conditionAsset;
+  transaction.conditionAmount = strategyPrep.conditionAmount;
+  transaction.timeCondition = strategyPrep.timeCondition;
+  transaction.timeConditionTZ = strategyPrep.timeConditionTZ;
+  transaction.strategyInstanceId = strategyPrep.instanceId;
+  transaction.strategyPrepId = strategyPrep.id;
+
+  return { transaction, isStrategyTx };
+}
+
+function decodeTxForStrategyPrep(transaction: IScheduled): IStrategyPrepTx {
+  switch (transaction.assetType) {
+    case AssetType.Ethereum:
+    case undefined: {
+      const parsed = ethers.utils.parseTransaction(transaction.signedTransaction);
+
+      return {
+        assetType: AssetType.Ethereum,
+        chainId: parsed.chainId,
+        from: parsed.from,
+        to: parsed.to,
+        nonce: parsed.nonce,
+        data: parsed.data,
+      };
+    }
+    case AssetType.Polkadot:
+    default: {
+      throw new Error('Implementme!');
```

Comment:
tiny typo Implement!

Reply:
It's actually supposed to be like this. It says 'implement me!', i.e. the case is asking to be implemented. It's just a convention